File: 백준/Silver/9012. 괄호/괄호.py
import sys
input = sys.stdin.readline

def check(string):
    stack=[]
    for c in string:
        if c=='(':
            stack.append(c)
        else:
            if stack:
                stack.pop()
            else:
                return False
    return len(stack)==0

for i in range(int(input())):
    string=input().strip()
    if check(string):
        print("YES")
    else:
        print("NO")

File: 백준/Silver/9012. 괄호/괄호.cc
#include <stdio.h>
#include <string.h>

#define MAX_LEN 51

int check(char *string) {
    int stack = 0;
    for (int i = 0; i < strlen(string); i++) {
        if (string[i] == '(') {
            stack++;
        } else if (string[i] == ')') {
            if (stack > 0) {
                stack--;
            } else {
                return 0;
            }
        }
    }
    return stack == 0;
}

int main() {
    int n;
    scanf("%d", &n);
    char string[MAX_LEN];

    for (int i = 0; i < n; i++) {
        scanf("%s", string);
        if (check(string)) {
            printf("YES\n");
        } else {
            printf("NO\n");
        }
    }

    return 0;
}